stl/inc/ranges

```diff
@@ -4749,6 +4752,400 @@ namespace ranges {
     } // namespace views
 
 #if _HAS_CXX23
+    template <class _Rng, class _Pat>
+    concept _Compatible_joinable_ranges = common_with<range_value_t<_Rng>, range_value_t<_Pat>> //
+        && common_reference_with<range_reference_t<_Rng>, range_reference_t<_Pat>> //
+        && common_reference_with<range_rvalue_reference_t<_Rng>, range_rvalue_reference_t<_Pat>>;
+
+    template <class _Rng>
+    concept _Bidirectional_common = bidirectional_range<_Rng> && common_range<_Rng>;
+
+    template <class _Rng, class _Pat> // TRANSITION, LLVM-47414
+    concept _Can_const_join_with =
+        input_range<const _Rng> && forward_range<const _Pat> && is_reference_v<range_reference_t<const _Rng>>;
+
+    template <input_range _Vw, forward_range _Pat>
+        requires view<_Vw> && input_range<range_reference_t<_Vw>> && view<_Pat> //
+            && _Compatible_joinable_ranges<range_reference_t<_Vw>, _Pat>
+    class join_with_view;
+
+    template <class _Vw, class _Pat>
+    class _Join_with_view_base : public view_interface<join_with_view<_Vw, _Pat>> {
+    private:
+        struct _Cache_wrapper {
+            template <input_iterator _Iter>
+            constexpr _Cache_wrapper(_Not_quite_object::_Construct_tag, const _Iter& _It) noexcept(noexcept(*_It))
+                : _Val(*_It) {}
+
+            remove_cv_t<range_reference_t<_Vw>> _Val;
+        };
+
+    protected:
+        /* [[no_unique_address]] */ _Non_propagating_cache<_Cache_wrapper, false> _Inner{};
+    };
+
+    template <class _Vw, class _Pat>
+        requires is_reference_v<range_reference_t<_Vw>>
+    class _Join_with_view_base<_Vw, _Pat> : public view_interface<join_with_view<_Vw, _Pat>> {
+    };
+
+    template <input_range _Vw, forward_range _Pat>
+        requires view<_Vw> && input_range<range_reference_t<_Vw>> && view<_Pat> //
+            && _Compatible_joinable_ranges<range_reference_t<_Vw>, _Pat>
+    class join_with_view : public _Join_with_view_base<_Vw, _Pat> {
+    private:
+        template <bool _Const>
+        using _InnerRng = range_reference_t<_Maybe_const<_Const, _Vw>>;
+
+        /* [[no_unique_address]] */ _Vw _Range{};
+        /* [[no_unique_address]] */ _Pat _Pattern{};
+
+        template <bool _Const>
+        class _Sentinel;
+
+        template <class _Outer, class _Inner, class _Pattern, bool _Deref_is_glvalue>
+        struct _Category_base {};
+
+        template <forward_range _Outer, forward_range _Inner, class _Pattern>
+        struct _Category_base<_Outer, _Inner, _Pattern, true> {

```

Comment:
Can we simplify that base type by only templatizing it as `template <bool _Const>` and then deducing all the stuff from that?